### PR TITLE
Improve nft

### PR DESCRIPTION
**Pull request by xlc**
Add class attributes
Add token attributes
Allow change class properties

TODOs:
- [x] tests
- [x] comments
- [x] weights & benchmarking

### CODE REVIEW

File: modules/nft/src/lib.rs [hunk 1]
```diff
@@ -51,14 +52,19 @@ pub use module::*;
...
 #[derive(Encode, Decode, Clone, Copy, BitFlags, RuntimeDebug, PartialEq, Eq)]
 pub enum ClassProperty {
-	/// Token can be transferred
+	/// Is token transferable
 	Transferable = 0b00000001,
-	/// Token can be burned
+	/// Is token burnable
 	Burnable = 0b00000010,
+	/// Allow mint new tokens
+	Mintable = 0b00000100,
```
Comment (zjb0807) [inline modules/nft/src/lib.rs:65]: unused?

**Comment by xlc** [comment]
./bench runtime module module_nft

File: modules/nft/src/lib.rs (hunk 1 above)
Comment (xlc) [reply modules/nft/src/lib.rs:65]: I missed it. Should be used now.

**Review by zjb0807 — APPROVED** [review]
LGTM. benchmarks and integrations need to update.

File: modules/nft/src/lib.rs [hunk 2]
```diff
@@ -51,14 +52,19 @@ pub use module::*;
...
 #[repr(u8)]
 #[derive(Encode, Decode, Clone, Copy, BitFlags, RuntimeDebug, PartialEq, Eq)]
 pub enum ClassProperty {
-	/// Token can be transferred
+	/// Is token transferable
 	Transferable = 0b00000001,
-	/// Token can be burned
+	/// Is token burnable
 	Burnable = 0b00000010,
+	/// Allow mint new tokens
```
Comment (shaunxw) [inline modules/nft/src/lib.rs:64]:
```suggestion
	/// Is minting new tokens allowed
```

File: modules/nft/src/lib.rs [hunk 3]
```diff
@@ -380,13 +449,36 @@ impl<T: Config> Pallet<T> {
...
+		}
+
 		Ok(())
 	}
+
+	fn data_deposit(metadata: &[u8], attributes: &Attributes) -> Result<BalanceOf<T>, DispatchError> {
+		// Addition can't overflow because we will be out of memory before that
+		let attributes_len = attributes
+			.iter()
+			.fold(0, |acc, (k, v)| (v.len() + k.len()) as u32 + acc);
```
Comment (shaunxw) [inline modules/nft/src/lib.rs:472]: `saturated` arithmetic might be better.

File: modules/nft/src/lib.rs [hunk 4]
```diff
@@ -380,13 +449,36 @@ impl<T: Config> Pallet<T> {
...
+
+		ensure!(
+			attributes_len <= T::MaxAttributesBytes::get(),
+			Error::<T>::AttributesTooLarge
+		);
+
+		let total_data_len = metadata.len() as u32 + attributes_len;
```
Comment (shaunxw) [inline modules/nft/src/lib.rs:479]: Same here

**Review by shaunxw — APPROVED** [review]
LGTM. Some nitpicks.

File: modules/nft/src/lib.rs [hunk 5]
```diff
@@ -83,17 +89,21 @@ impl Decode for Properties {
...
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 pub struct ClassData<Balance> {
-	/// The minimum balance to create class
+	/// Deposit reserved to create token class
 	pub deposit: Balance,
-	/// Property of token
+	/// Class properties
 	pub properties: Properties,
+	/// Class attributes
+	pub attributes: Attributes,
```
Comment (xlc) [inline modules/nft/src/lib.rs:97]: requires acala.js update @qwer951123 @ntduan 
`ClassProperty`, `ClassData`, `TokenData`

**Pull request merged**